<a href="https://colab.research.google.com/github/douglascdsantos/data_science/blob/main/dados_ausentes/notebooks/lidando_com_dados_ausentes_com_dados_de_ocorrencias_de_aviacao.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd


In [ ]:
url = "https://raw.githubusercontent.com/carlosfab/dsnp2/master/datasets/ocorrencias_aviacao.csv"
df = pd.read_csv(url,
                 index_col = 'codigo_ocorrencia')

In [ ]:
df.head()

ocorrencia_classificacao  ... ocorrencia_dia_extracao
codigo_ocorrencia                           ...                        
201305055424986                   ACIDENTE  ...              2018-08-28
201605160250139            INCIDENTE GRAVE  ...              2018-08-28
201805021421302                  INCIDENTE  ...              2018-08-28
201311259977425            INCIDENTE GRAVE  ...              2018-08-28
201103187273112                  INCIDENTE  ...              2018-08-28

[5 rows x 21 columns]

Conferindo as features do nosso dataset, sabemos que é esperado de que as colunas
- ocorrencia_dia;
- ocorrencia_horario;
- divulgacao_dia_publicacao	;
- ocorrencia_dia_extracao.

venham com dtype de *datetime*

In [ ]:
# conferindo os dtypes do dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5208 entries, 201305055424986 to 201309012098180
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ocorrencia_classificacao        5208 non-null   object 
 1   ocorrencia_tipo                 5208 non-null   object 
 2   ocorrencia_tipo_categoria       5208 non-null   object 
 3   ocorrencia_tipo_icao            5208 non-null   object 
 4   ocorrencia_latitude             5208 non-null   float64
 5   ocorrencia_longitude            5208 non-null   float64
 6   ocorrencia_cidade               5208 non-null   object 
 7   ocorrencia_uf                   5208 non-null   object 
 8   ocorrencia_pais                 5208 non-null   object 
 9   ocorrencia_aerodromo            5208 non-null   object 
 10  ocorrencia_dia                  5208 non-null   object 
 11  ocorrencia_horario              5208 non-null   object 
 12  investiga

Como as features citadas não estavam, vamos transforma-las para o formato esperado

In [ ]:
# transformando a coluna 'ocorrencia_dia' e 'ocorrencia_horario' em datetime
df['ocorrencia_dia'] = pd.to_datetime(df['ocorrencia_dia'])
df['ocorrencia_horario'] = pd.to_datetime(df['ocorrencia_horario'])
df['divulgacao_dia_publicacao'] = pd.to_datetime(df['divulgacao_dia_publicacao'])
df['ocorrencia_dia_extracao'] = pd.to_datetime(df['ocorrencia_dia_extracao'])

# como a feature 'ocorrencia_horario' possui somente o horário se rodarmos somente o código acima irá retornar preenchendo o dia com o dia de hoje ficando do tipo '2021-10-14 19:19:00'
df['ocorrencia_horario'] = df['ocorrencia_horario'].dt.strftime('%H:%M:%S')

In [ ]:
# conferindo a alteração
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5208 entries, 201305055424986 to 201309012098180
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   ocorrencia_classificacao        5208 non-null   object        
 1   ocorrencia_tipo                 5208 non-null   object        
 2   ocorrencia_tipo_categoria       5208 non-null   object        
 3   ocorrencia_tipo_icao            5208 non-null   object        
 4   ocorrencia_latitude             5208 non-null   float64       
 5   ocorrencia_longitude            5208 non-null   float64       
 6   ocorrencia_cidade               5208 non-null   object        
 7   ocorrencia_uf                   5208 non-null   object        
 8   ocorrencia_pais                 5208 non-null   object        
 9   ocorrencia_aerodromo            5208 non-null   object        
 10  ocorrencia_dia                  5208 non-null  

In [ ]:
df.head()

ocorrencia_classificacao  ... ocorrencia_dia_extracao
codigo_ocorrencia                           ...                        
201305055424986                   ACIDENTE  ...              2018-08-28
201605160250139            INCIDENTE GRAVE  ...              2018-08-28
201805021421302                  INCIDENTE  ...              2018-08-28
201311259977425            INCIDENTE GRAVE  ...              2018-08-28
201103187273112                  INCIDENTE  ...              2018-08-28

[5 rows x 21 columns]

Já ajustado os dtypes, vamos olhar para os valores não nulos do nosso dataset, sabemos que temos 5208 entradas, então, na coluna 'Non-Null Count' toda entrada que não for 5208 deve ser analisada

Aqui temos duas features do nosso dataset para analisarmos
- investigacao_aeronave_liberada
- divulgacao_dia_publicacao

Começando com a primeira:

In [ ]:
# distribuição dos dados não nulos da coluna 'investigacao_aeronave_liberada'
df.investigacao_aeronave_liberada[df.investigacao_aeronave_liberada.notnull()].value_counts()

SIM    1020
NÃO      81
***       9
Name: investigacao_aeronave_liberada, dtype: int64

In [ ]:
# olhando como percentual, temos
df.investigacao_aeronave_liberada.value_counts(normalize=True,) * 100

SIM    91.891892
NÃO     7.297297
***     0.810811
Name: investigacao_aeronave_liberada, dtype: float64

In [ ]:
# quantidade de dados aparentemente nulos na coluna 'investigacao_aeronave_liberada'
df.investigacao_aeronave_liberada.isnull().value_counts()

True     4098
False    1110
Name: investigacao_aeronave_liberada, dtype: int64

In [ ]:
# considerando de que é esperado que essa coluna seja um dado categorio de 'Sim' ou 'Não', vamos preencher o valores NaN com 'Não', talvez o sistema de contabilização seja de um checklist e o NaN são de quando não foi o 'SIM'
df.investigacao_aeronave_liberada.fillna('NÃO',inplace=True)

In [ ]:
# Olhando novamente distribuição dos dados não nulos da coluna 'investigacao_aeronave_liberada', desta vez com os dados faltantes preenchidos
df.investigacao_aeronave_liberada[df.investigacao_aeronave_liberada.notnull()].value_counts()

NÃO    4179
SIM    1020
***       9
Name: investigacao_aeronave_liberada, dtype: int64

In [ ]:
# Novamnete, olhando como percentual, temos
df.investigacao_aeronave_liberada.value_counts(normalize=True,) * 100

NÃO    80.241935
SIM    19.585253
***     0.172811
Name: investigacao_aeronave_liberada, dtype: float64

A conclusão a primeira vista, sem tratar os dados faltantes, estariamos inclicados a pensar de que a maior parte das ocorrencias do nosso dataset se deu com a investigação da aeronave liberada, quando na verdade não estava.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5208 entries, 201305055424986 to 201309012098180
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   ocorrencia_classificacao        5208 non-null   object        
 1   ocorrencia_tipo                 5208 non-null   object        
 2   ocorrencia_tipo_categoria       5208 non-null   object        
 3   ocorrencia_tipo_icao            5208 non-null   object        
 4   ocorrencia_latitude             5208 non-null   float64       
 5   ocorrencia_longitude            5208 non-null   float64       
 6   ocorrencia_cidade               5208 non-null   object        
 7   ocorrencia_uf                   5208 non-null   object        
 8   ocorrencia_pais                 5208 non-null   object        
 9   ocorrencia_aerodromo            5208 non-null   object        
 10  ocorrencia_dia                  5208 non-null  

Dessa vez tratando a segunda (divulgacao_dia_publicacao), vamos verificar o percentual de dados faltantes para então decidirmos se vamos excluir os dados (caso seja um valor muito pequeno do dataframe) ou se vamos encontrar alguma maneira de preenche-los (quando a quantidade de dados ausentes for significativa)

In [ ]:
# verificando a o percentual de dados faltantes
df.divulgacao_dia_publicacao.isnull().value_counts(normalize=True) * 100

True     75.78725
False    24.21275
Name: divulgacao_dia_publicacao, dtype: float64

Como os dados faltantes representam >75% do nosso dataset vamos preenche-lo, a maneira encontrada de fazer isso foi:

Considerando os dados não nulos, foi calculado o tempo médio entre a data de ocorrencia e a de publicação então foi somado a data de ocorrencia dos valores nulos preenchendo assim os campos nulos na feature 'divulgacao_dia_publicacao'.

In [ ]:
# cálculo da média a ser utilizada
dias_entre_ocorrencia_e_publicacao = df[df.divulgacao_dia_publicacao.notnull()].divulgacao_dia_publicacao - df[df.divulgacao_dia_publicacao.notnull()].ocorrencia_dia
dias_entre_ocorrencia_e_publicacao.mean()

Timedelta('767 days 21:09:50.959555912')

In [ ]:
# para fins de comparação, o codigo de ocorencia
df[df.divulgacao_dia_publicacao.notnull()].ocorrencia_dia

codigo_ocorrencia
201311259977425   2013-11-25
200808073991179   2008-08-07
201108118145873   2011-08-11
201604062021424   2016-04-05
201208018273083   2012-08-01
                     ...    
201412032134334   2014-12-03
201505293589133   2015-05-29
201406236453857   2014-06-23
201704150134378   2017-04-12
201309012098180   2013-09-02
Name: ocorrencia_dia, Length: 1261, dtype: datetime64[ns]

Afim de verificar de uma maneira visual é esperado que cada linhas a baixo seja a linha de cima acrescentada de 767 dias(aproximadamente 2 anos)

In [ ]:
# código de ocorrencia acrescentada de 767 dias
df[df.divulgacao_dia_publicacao.notnull()].ocorrencia_dia + dias_entre_ocorrencia_e_publicacao.mean()

codigo_ocorrencia
201311259977425   2016-01-01 21:09:50.959555912
200808073991179   2010-09-13 21:09:50.959555912
201108118145873   2013-09-16 21:09:50.959555912
201604062021424   2018-05-12 21:09:50.959555912
201208018273083   2014-09-07 21:09:50.959555912
                               ...             
201412032134334   2017-01-08 21:09:50.959555912
201505293589133   2017-07-04 21:09:50.959555912
201406236453857   2016-07-29 21:09:50.959555912
201704150134378   2019-05-19 21:09:50.959555912
201309012098180   2015-10-09 21:09:50.959555912
Name: ocorrencia_dia, Length: 1261, dtype: datetime64[ns]

In [ ]:
# Visualização do index da seleção acima
df.divulgacao_dia_publicacao.isnull().index

Int64Index([201305055424986, 201605160250139, 201805021421302, 201311259977425,
            201103187273112, 200808073991179, 201108118145873, 201604251335501,
            201604062021424, 200902274965349,
            ...
            200810155397991, 201502247981603, 200811129767665, 200811224318981,
            200811226332275, 200811264169752, 200910311058203, 200810318709738,
            200803063018556, 201309012098180],
           dtype='int64', name='codigo_ocorrencia', length=5208)

In [ ]:
# Preenchendo somente os valores vazios em df utilizando do laço de repetição for
for i in df.divulgacao_dia_publicacao.isnull().index:
  df.loc[i,'divulgacao_dia_publicacao'] = df.loc[i,'ocorrencia_dia'] + dias_entre_ocorrencia_e_publicacao.mean()

In [ ]:
df.divulgacao_dia_publicacao

codigo_ocorrencia
201305055424986   2015-06-11 21:09:50.959555912
201605160250139   2018-06-19 21:09:50.959555912
201805021421302   2020-04-14 21:09:50.959555912
201311259977425   2016-01-01 21:09:50.959555912
201103187273112   2013-04-23 21:09:50.959555912
                               ...             
200811264169752   2011-01-02 21:09:50.959555912
200910311058203   2011-12-07 21:09:50.959555912
200810318709738   2010-12-07 21:09:50.959555912
200803063018556   2010-04-12 21:09:50.959555912
201309012098180   2015-10-09 21:09:50.959555912
Name: divulgacao_dia_publicacao, Length: 5208, dtype: datetime64[ns]

In [ ]:
# como visto acima não está mostrando somente o dia como é esperado, dessa forma precisamos formatar para ano-mês-dia
df['divulgacao_dia_publicacao'] = df['divulgacao_dia_publicacao'].dt.strftime('%Y-%m-%d')
# Nesta formatação acabou convertando o dtype de datetime para string, então, convertendo novamente
df['divulgacao_dia_publicacao'] = pd.to_datetime(df['divulgacao_dia_publicacao'])

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5208 entries, 201305055424986 to 201309012098180
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   ocorrencia_classificacao        5208 non-null   object        
 1   ocorrencia_tipo                 5208 non-null   object        
 2   ocorrencia_tipo_categoria       5208 non-null   object        
 3   ocorrencia_tipo_icao            5208 non-null   object        
 4   ocorrencia_latitude             5208 non-null   float64       
 5   ocorrencia_longitude            5208 non-null   float64       
 6   ocorrencia_cidade               5208 non-null   object        
 7   ocorrencia_uf                   5208 non-null   object        
 8   ocorrencia_pais                 5208 non-null   object        
 9   ocorrencia_aerodromo            5208 non-null   object        
 10  ocorrencia_dia                  5208 non-null  

In [ ]:
# verificando a existencia de algum dado ausente
df.isnull().sum().sum()

0

Nesta parte já temos o nosso dataset sem dados faltantes e com os dtypes corretos